File: chrome/test/data/extensions/api_test/automation/tests/tabs/location.js
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

var allTests = [
  function testLocation() {
    function assertOkButtonLocation(event) {
      var okButton = rootNode.firstChild().firstChild();
      assertTrue('location' in okButton);
      assertEq({left:100, top: 200, width: 300, height: 400},
               okButton.location);
      chrome.test.succeed();
    };

    var okButton = rootNode.firstChild().firstChild();
    assertTrue('location' in okButton, 'no location in okButton');
    assertTrue('left' in okButton.location, 'no left in location');
    assertTrue('top' in okButton.location, 'no top in location');
    assertTrue('height' in okButton.location, 'no height in location');
    assertTrue('width' in okButton.location, 'no width in location');

    rootNode.addEventListener(
        EventType.childrenChanged, assertOkButtonLocation);
    chrome.tabs.executeScript({ 'code':
          'document.querySelector("button")' +
          '.setAttribute("style", "position: absolute; left: 100; top: 200; ' +
          'width: 300; height: 400;");' });
  }
];

setUpAndRunTests(allTests);

File: tools/gn/location.cc
// Copyright (c) 2013 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "tools/gn/location.h"

#include "base/logging.h"
#include "base/strings/string_number_conversions.h"
#include "tools/gn/input_file.h"

Location::Location()
    : file_(NULL),
      line_number_(-1),
      char_offset_(-1) {
}

Location::Location(const InputFile* file,
                   int line_number,
                   int char_offset,
                   int byte)
    : file_(file),
      line_number_(line_number),
      char_offset_(char_offset),
      byte_(byte) {
}

bool Location::operator==(const Location& other) const {
  return other.file_ == file_ &&
         other.line_number_ == line_number_ &&
         other.char_offset_ == char_offset_;
}

bool Location::operator!=(const Location& other) const {
  return !operator==(other);
}

bool Location::operator<(const Location& other) const {
  DCHECK(file_ == other.file_);
  if (line_number_ != other.line_number_)
    return line_number_ < other.line_number_;
  return char_offset_ < other.char_offset_;
}

std::string Location::Describe(bool include_char_offset) const {
  if (!file_)
    return std::string();

  std::string ret;
  if (file_->friendly_name().empty())
    ret = file_->name().value();
  else
    ret = file_->friendly_name();

  ret += ":";
  ret += base::IntToString(line_number_);
  if (include_char_offset) {
    ret += ":";
    ret += base::IntToString(char_offset_);
  }
  return ret;
}

LocationRange::LocationRange() {
}

LocationRange::LocationRange(const Location& begin, const Location& end)
    : begin_(begin),
      end_(end) {
  DCHECK(begin_.file() == end_.file());
}

LocationRange LocationRange::Union(const LocationRange& other) const {
  DCHECK(begin_.file() == other.begin_.file());
  return LocationRange(
      begin_ < other.begin_ ? begin_ : other.begin_,
      end_ < other.end_ ? other.end_ : end_);
}
